devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy

```diff
@@ -0,0 +1,480 @@
+		optionValue = getConfigurationInfo(1, "enumMap").find { it.key == state."${getConfigurationInfo(1, "name")}" }?.value
+		cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x01, scale: optionValue?:0x00).format()
+	}
+	if (device?.currentValue("illuminance") == null) {
+		cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x03, scale: 0x00).format()
+	}
+	if (device?.currentValue("humidity") == null) {
+		cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x05, scale: 0x00).format()
+	}
+	if (canReportBattery() || device?.currentValue("battery") == null) {
+		cmds << zwave.batteryV1.batteryGet().format()
+	}
+	
+	getConfigurationNumber().each { idx ->
+		if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
+			switch (getConfigurationInfo(idx, "type")) {
+				case "enum":
+					optionValue = getConfigurationInfo(idx, "enumMap").find { it.key == state."${getConfigurationInfo(idx, "name")}" }?.value
+					value = integer2Array(optionValue, getConfigurationInfo(idx, "size"))
+					break
+				case "number":
+					value = integer2Array(state."${getConfigurationInfo(idx, "name")}", getConfigurationInfo(idx, "size"))
+					break
+			}
+			if (value != null) {
+				cmds << zwave.configurationV2.configurationSet(parameterNumber: idx, defaultValue: false, size: getConfigurationInfo(idx, "size"), configurationValue: value).format()
+				cmds << zwave.configurationV2.configurationGet(parameterNumber: idx).format()
+				if (idx == 1) {
+					cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x01, scale: optionValue?:0x00).format()
+				}
+				value = null
+			}
+		}
+	}
+	
+	if (state."${getWakeUpInfo("name")}Refresh" == true) {
+		 cmds << zwave.wakeUpV2.wakeUpIntervalSet(nodeid: zwaveHubNodeId, seconds: hour2Second(state.wakeUpInterval)).format()
+		 cmds << zwave.wakeUpV2.wakeUpIntervalGet().format()
+	}
+	
+	sendEvent(name: "checkInterval", value: (hour2Second(state.wakeUpInterval) + 2 * 60) * 2, displayed: false, data: [protocol: "zwave", hubHardwareId: device.hub.hardwareID])
+	
+	return cmds ? delayBetween(cmds, 500) : []
+}
+
+def configurationUpdate(config) {
+	getConfigurationNumber().each { idx ->
+		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
+			if (state."${getConfigurationInfo(idx, "name")}" != config?."${getConfigurationInfo(idx, "name")}") {
+				state."${getConfigurationInfo(idx, "name")}" = config?."${getConfigurationInfo(idx, "name")}"
+				state."${getConfigurationInfo(idx, "name")}Refresh" = true
+			}
+		}
+	}
+}
+
+def configurationCheck(config) {
+	getConfigurationNumber().each { idx ->
+		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
+			if (config?."${getConfigurationInfo(idx, "name")}" == state."${getConfigurationInfo(idx, "name")}") {
+				if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
+					state."${getConfigurationInfo(idx, "name")}Refresh" = false
+				}
+			}
+		}
+	}
+}
```

Comment:
if you wrote your configuration report hander function differently and refactored your configuration parameters as a map of maps, I think this method could instead be written:

```
config.each { state[key].Refresh = false}
```

where `state[x]` is a map containing a value with key `value` and a boolean with key `Refresh` and your configuration report handler returns a list of numbers.

Because as far as I can tell you're not actually comparing the value you've received from the configuration report to the value the user has set. Any configuration report indicating a parameter number the user has set will get marked as not needing to be refreshed, regardless of the actual value reported.

This is just one example here. I'm not really satisfied with how you've handled configuration options. For a more acceptable example take a look at any of the Fibaro DTHs that include preferences.

Reply:
> Because as far as I can tell you're not actually comparing the value you've received from the configuration report to the value the user has set. Any configuration report indicating a parameter number the user has set will get marked as not needing to be refreshed, regardless of the actual value reported.

Actually, I will compare the value I am received from the configuration report to the value the user has set. If these values are different, I will send the configuration value the user has set on next time I received the wake up notification until these value are the same. 

Here is my configuration snippet code, and its procedure. Please look this, and give me a suggestion, thanks.

Variable definition:
1. _**state."${getConfigurationInfo(idx, "name")}"**_: The values the user has set.
2. _**state."${getConfigurationInfo(idx,"name")}Refresh"**_:  This configuration is needed to be sent or not: true is needed to be sent while I received the wake up notification, false is not.

Procedure:
1. After the user changed the configuration value, I will check which configuration value is changed. If this configuration value has changed, I will store this configuration value to corresponding _**state."${getConfigurationInfo(idx, "name")}"**_, and set _**state."${getConfigurationInfo(idx, "name")}Refresh"**_ to true to indicate this needs to be updated to this device. Here are **updated()** and **configurationUpdate(settings)**.
2. I will send this configuration set with _**state."${getConfigurationInfo(idx, "name")}"**_ if its _**state."${getConfigurationInfo(idx, "name")}Refresh"**_ is true after I received the wake up notification. And then I will send the configuration get to check the configuration report value is the same or not. Here are z**waveEvent(physicalgraph.zwave.commands.wakeupv2.WakeUpNotification cmd)** and **configure()**.
3. I will compare the configuration report I received to _**state."${getConfigurationInfo(idx, "name")}"**_ from the user has set. I will set _**state."${getConfigurationInfo(idx, "name")}Refresh"**_ to false if these are the same. Here are **zwaveEvent(physicalgraph.zwave.commands.configurationv2.ConfigurationReport cmd)** and **configurationCheck(config)**.
4. Loop from step 2 to step 3 until these configuration value are the same.

```
def updated() {
	configurationUpdate(settings)
	...
}

def configurationUpdate(config) {
	getConfigurationNumber().each { idx ->
		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
			if (state."${getConfigurationInfo(idx, "name")}" != config?."${getConfigurationInfo(idx, "name")}") {
				state."${getConfigurationInfo(idx, "name")}" = config?."${getConfigurationInfo(idx, "name")}"
				state."${getConfigurationInfo(idx, "name")}Refresh" = true
			}
		}
	}
}

def zwaveEvent(physicalgraph.zwave.commands.wakeupv2.WakeUpNotification cmd) {
	def cmds = []
	
	cmds += configure()
	if (cmds) {
		cmds << "delay 5000"
	}
	cmds << zwave.wakeUpV2.wakeUpNoMoreInformation().format()
	
	return cmds ? response(cmds) : []
}

def configure() {
	def cmds = []
	def value
	def optionValue
	...
	getConfigurationNumber().each { idx ->
		if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
			switch (getConfigurationInfo(idx, "type")) {
				case "enum":
					optionValue = getConfigurationInfo(idx, "enumMap").find { it.key == state."${getConfigurationInfo(idx, "name")}" }?.value
					value = integer2Array(optionValue, getConfigurationInfo(idx, "size"))
					break
				case "number":
					value = integer2Array(state."${getConfigurationInfo(idx, "name")}", getConfigurationInfo(idx, "size"))
					break
			}
			if (value != null) {
				cmds << zwave.configurationV2.configurationSet(parameterNumber: idx, defaultValue: false, size: getConfigurationInfo(idx, "size"), configurationValue: value).format()
				cmds << zwave.configurationV2.configurationGet(parameterNumber: idx).format()
				if (idx == 1) {
					cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x01, scale: optionValue?:0x00).format()
				}
				value = null
			}
		}
	}
	...
	return cmds ? delayBetween(cmds, 500) : []
}

def zwaveEvent(physicalgraph.zwave.commands.configurationv2.ConfigurationReport cmd) {
	def size = getConfigurationInfo(cmd.parameterNumber, "size")
	def value = array2Integer(cmd.configurationValue)
	def config = [:]
	
	if (size && size == cmd.size) {
		switch (getConfigurationInfo(cmd.parameterNumber, "type")) {
			case "enum":
				def optionName = getConfigurationInfo(cmd.parameterNumber, "enumMap").find { it.value == value}?.key
				if (optionName) {
					config."${getConfigurationInfo(cmd.parameterNumber, "name")}" = optionName
				}
				break
			case "number":
				config."${getConfigurationInfo(cmd.parameterNumber, "name")}" = value
				break
		}
		
		if (config) {
			configurationCheck(config)
		}
	}
	[]
}

def configurationCheck(config) {
	getConfigurationNumber().each { idx ->
		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
			if (config?."${getConfigurationInfo(idx, "name")}" == state."${getConfigurationInfo(idx, "name")}") {
				if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
					state."${getConfigurationInfo(idx, "name")}Refresh" = false
				}
			}
		}
	}
}
```

Reply:
Okay so you are verifying they've been set.

My mistake, but this is just reinforcing my point that the code you've written is hard to follow because you're using integers combined with strings as your dictionary keys.

Iterating from 1..8 and checking to see if each param was in your updated configuration report is much more laborious and confusing than just structuring the output of your configuration report such that you could simply iterate over that itself.

There are other existing DTHs that solve this problem more elegantly: https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/fibargroup/fibaro-dimmer-2-zw5.src/fibaro-dimmer-2-zw5.groovy for example.

Reply:
I think I understand your meaning. Your approach simplifies this procedure and make it more elegantly.
I refactored configuration parameter using a map of maps, and refactored wake up info parameter using a map. Please review it. Thank you very much.